arch/RISC-V/src/rv32/riscv-csr_generic-read_csr_64.adb

```diff
@@ -0,0 +1,40 @@
+------------------------------------------------------------------------------
+--                                                                          --
+--                        Copyright (C) 2020, AdaCore                       --
+--                                                                          --
+--  Redistribution and use in source and binary forms, with or without      --
+--  modification, are permitted provided that the following conditions are  --
+--  met:                                                                    --
+--     1. Redistributions of source code must retain the above copyright    --
+--        notice, this list of conditions and the following disclaimer.     --
+--     2. Redistributions in binary form must reproduce the above copyright --
+--        notice, this list of conditions and the following disclaimer in   --
+--        the documentation and/or other materials provided with the        --
+--        distribution.                                                     --
+--     3. Neither the name of the copyright holder nor the names of its     --
+--        contributors may be used to endorse or promote products derived   --
+--        from this software without specific prior written permission.     --
+--                                                                          --
+--   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS    --
+--   "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT      --
+--   LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR  --
+--   A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT   --
+--   HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, --
+--   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT       --
+--   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,  --
+--   DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY  --
+--   THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT    --
+--   (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE  --
+--   OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.   --
+--                                                                          --
+------------------------------------------------------------------------------
+
+with HAL; use HAL;
+
+separate (RISCV.CSR_Generic)
+function Read_CSR_64 return HAL.UInt64 is
+   function Low is new Read_CSR (Reg_Name, UInt32);
+   function High is new Read_CSR (Reg_Name & "h", UInt32);
+begin
+   return Shift_Left(UInt64 (High), 32) + UInt64 (Low);
```

Comment:
Actually... only the "volatile" counter CSRs need this special treatment - so mcycle, minstret, mtime, and the mhpmcounter* CSRs. I assume that all the others are not considered volatile so OK to read in the simpler fashion. In my implementation, the function specification for getting MCycle is:
```
   function Cycle return UInt64
     with Import,
          Convention => Assembler,
          Link_Name => "_rv32_read_cycle";
```
and implemented in Assembler as:
```
        .globl _rv32_read_cycle
_rv32_read_cycle:
	/* Note that we loop here to check the msb doesn't change   */
	/* while reading the lsb. See RISC-V ISA Specification 10.1 */
        rdcycleh t0
        rdcycle  t1
        rdcycleh t2
        bne t0, t2, _rv32_read_cycle
	mv a0, t1
	mv a1, t2
	ret
```
Note that I use t0, t1 and t2 as temporary registers since these do not need to be saved and restored across the function call.

Reply:
Right, I will fix that.